File: kalitech-engine/src/main/resources/kalitech/engine/modules/Terrain/helpers/TerrainPhysics.js
// FILE: resources/kalitech/builtin/terrain/TerrainPhysics.js
"use strict";

const {isObj} = require("./TerrainTypes.js");

function surfaceIdOf(h) {
    if (typeof h === "number") return h | 0;
    if (!h) return 0;
    if (typeof h.id === "function") return h.id() | 0;
    if (typeof h.id === "number") return h.id | 0;
    if (typeof h.surfaceId === "number") return h.surfaceId | 0;
    return 0;
}

function bodyIdOf(h) {
    if (typeof h === "number") return h | 0;
    if (!h) return 0;
    if (typeof h.id === "function") return h.id() | 0;
    if (typeof h.id === "number") return h.id | 0;
    if (typeof h.bodyId === "number") return h.bodyId | 0;
    return 0;
}

class TerrainPhysics {
    constructor(engine) {
        this.engine = engine;
    }

    resolveBodyId(surfaceHandleOrId, maybeBodyHandleOrId) {
        const sid = surfaceIdOf(surfaceHandleOrId);
        if (sid <= 0) return 0;

        const E = this.engine;

        // prefer surface-attached body (surface API)
        if (E.surface && typeof E.surface === "function") {
            const s = E.surface();
            if (s && typeof s.attachedBody === "function") {
                const bid = bodyIdOf(s.attachedBody(sid));
                if (bid > 0) return bid;
            }
        }

        // fallback physics link (physics API)
        if (E.physics && typeof E.physics === "function") {
            const p = E.physics();
            if (p && typeof p.bodyOfSurface === "function") {
                const bid = bodyIdOf(p.bodyOfSurface(sid));
                if (bid > 0) return bid;
            }
        }

        const bid = bodyIdOf(maybeBodyHandleOrId);
        return (bid > 0) ? bid : 0;
    }

    ensureStaticBody(surfaceHandleOrId, physCfg, defaultColliderType) {
        const sid = surfaceIdOf(surfaceHandleOrId);
        if (sid <= 0) return {bodyId: 0, bodyHandle: null};

        const existing = this.resolveBodyId(sid, null);
        if (existing > 0) return {bodyId: existing, bodyHandle: null};

        const base = {
            surface: sid,
            mass: 0,
            kinematic: true,
            collider: {type: defaultColliderType || "mesh"},
        };
        const cfg = isObj(physCfg) ? Object.assign(base, physCfg) : base;

        // strict: physics must exist
        let bodyHandle = null;
        if (typeof PHYS !== "undefined" && PHYS && typeof PHYS.body === "function") {
            bodyHandle = PHYS.body(cfg);
        } else {
            const E = this.engine;
            const p = (E.physics && typeof E.physics === "function") ? E.physics() : null;
            if (!p || typeof p.body !== "function") throw new Error("[TERR] physics.body(cfg) is required");
            bodyHandle = p.body(cfg);
        }

        const bodyId = this.resolveBodyId(sid, bodyHandle);
        return {bodyId, bodyHandle};
    }

    withBody(terrNative, surface, physCfg, defaultColliderType) {
        if (physCfg == null) return surface;

        let bodyHandle = null;

        // prefer native terr.physics(surface,cfg) if present
        if (terrNative && typeof terrNative.physics === "function") {
            bodyHandle = terrNative.physics(surface, physCfg);
        }

        const sid = surfaceIdOf(surface);
        let bodyId = this.resolveBodyId(sid, bodyHandle);

        if (bodyId <= 0) {
            const made = this.ensureStaticBody(surface, physCfg, defaultColliderType || "mesh");
            bodyId = made.bodyId;
            if (!bodyHandle) bodyHandle = made.bodyHandle;
        }

        const bodyRef =
            (bodyId > 0 && typeof PHYS !== "undefined" && PHYS && typeof PHYS.ref === "function")
                ? PHYS.ref(bodyId)
                : undefined;

        return Object.freeze({surface, bodyId, body: bodyRef, handle: bodyHandle});
    }
}

module.exports = {TerrainPhysics, surfaceIdOf, bodyIdOf};

File: modulesSrc/TerrainModule/src/main/resources/kalitech/engine/modules/Terrain/helpers/TerrainPhysics.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local Tables = luaRuntime.table
local Classes = luaRuntime.class
local Error = luaRuntime.Error
local lua_require_result_0 = require("./TerrainTypes.lua")
isObj = lua_require_result_0.isObj
function surfaceIdOf(self, h)
    if KTypeOf(h) == "number" then
        return bit32.bor(h, 0)
    end
    if not h then
        return 0
    end
    if KTypeOf(h.id) == "function" then
        return bit32.bor(
            h:id(),
            0
        )
    end
    if KTypeOf(h.id) == "number" then
        return bit32.bor(h.id, 0)
    end
    if KTypeOf(h.surfaceId) == "number" then
        return bit32.bor(h.surfaceId, 0)
    end
    return 0
end
function bodyIdOf(self, h)
    if KTypeOf(h) == "number" then
        return bit32.bor(h, 0)
    end
    if not h then
        return 0
    end
    if KTypeOf(h.id) == "function" then
        return bit32.bor(
            h:id(),
            0
        )
    end
    if KTypeOf(h.id) == "number" then
        return bit32.bor(h.id, 0)
    end
    if KTypeOf(h.bodyId) == "number" then
        return bit32.bor(h.bodyId, 0)
    end
    return 0
end
TerrainPhysics = Classes:create()
TerrainPhysics.name = "TerrainPhysics"
function TerrainPhysics.prototype.lua_constructor(self, engine)
    self.engine = engine
end
function TerrainPhysics.prototype.resolveBodyId(self, surfaceHandleOrId, maybeBodyHandleOrId)
    local sid = surfaceIdOf(_G, surfaceHandleOrId)
    if sid <= 0 then
        return 0
    end
    local E = self.engine
    if E.surface and KTypeOf(E.surface) == "function" then
        local s = E:surface()
        if s and KTypeOf(s.attachedBody) == "function" then
            local bid = bodyIdOf(
                _G,
                s:attachedBody(sid)
            )
            if bid > 0 then
                return bid
            end
        end
    end
    if E.physics and KTypeOf(E.physics) == "function" then
        local p = E:physics()
        if p and KTypeOf(p.bodyOfSurface) == "function" then
            local bid = bodyIdOf(
                _G,
                p:bodyOfSurface(sid)
            )
            if bid > 0 then
                return bid
            end
        end
    end
    local bid = bodyIdOf(_G, maybeBodyHandleOrId)
    local lua_temp_1
    if bid > 0 then
        lua_temp_1 = bid
    else
        lua_temp_1 = 0
    end
    return lua_temp_1
end
function TerrainPhysics.prototype.ensureStaticBody(self, surfaceHandleOrId, physCfg, defaultColliderType)
    local sid = surfaceIdOf(_G, surfaceHandleOrId)
    if sid <= 0 then
        return {bodyId = 0, bodyHandle = nil}
    end
    local existing = self:resolveBodyId(sid, nil)
    if existing > 0 then
        return {bodyId = existing, bodyHandle = nil}
    end
    local base = {surface = sid, mass = 0, kinematic = true, collider = {type = defaultColliderType or "mesh"}}
    local lua_isObj_result_2
    if isObj(_G, physCfg) then
        lua_isObj_result_2 = Tables:merge(base, physCfg)
    else
        lua_isObj_result_2 = base
    end
    local cfg = lua_isObj_result_2
    local bodyHandle = nil
    if KTypeOf(ENGINE.physics) ~= "nil" and ENGINE.physics and KTypeOf(ENGINE.physics.body) == "function" then
        bodyHandle = ENGINE.physics:body(cfg)
    else
        local E = self.engine
        local lua_temp_3
        if E.physics and KTypeOf(E.physics) == "function" then
            lua_temp_3 = E:physics()
        else
            lua_temp_3 = nil
        end
        local p = lua_temp_3
        if not p or KTypeOf(p.body) ~= "function" then
            error(
                Classes:construct(Error, "[TERR] physics.body(cfg) is required"),
                0
            )
        end
        bodyHandle = p:body(cfg)
    end
    local bodyId = self:resolveBodyId(sid, bodyHandle)
    return {bodyId = bodyId, bodyHandle = bodyHandle}
end
function TerrainPhysics.prototype.withBody(self, terrNative, surface, physCfg, defaultColliderType)
    if physCfg == nil then
        return surface
    end
    local bodyHandle = nil
    if terrNative and KTypeOf(terrNative.physics) == "function" then
        bodyHandle = terrNative:physics(surface, physCfg)
    end
    local sid = surfaceIdOf(_G, surface)
    local bodyId = self:resolveBodyId(sid, bodyHandle)
    if bodyId <= 0 then
        local made = self:ensureStaticBody(surface, physCfg, defaultColliderType or "mesh")
        bodyId = made.bodyId
        if not bodyHandle then
            bodyHandle = made.bodyHandle
        end
    end
    local lua_temp_4
    if bodyId > 0 and KTypeOf(ENGINE.physics) ~= "nil" and ENGINE.physics and KTypeOf(ENGINE.physics.ref) == "function" then
        lua_temp_4 = ENGINE.physics:ref(bodyId)
    else
        lua_temp_4 = nil
    end
    local bodyRef = lua_temp_4
    return KObject:freeze({surface = surface, bodyId = bodyId, body = bodyRef, handle = bodyHandle})
end
M = {TerrainPhysics = TerrainPhysics, surfaceIdOf = surfaceIdOf, bodyIdOf = bodyIdOf}

return M
